### ISSUE TITLE
Filterable diagnostic names should be checked, but also permit third-party diagnostics

### ISSUE THREAD

**Issue by jimblandy** [gpuweb/gpuweb#3989]
Mistyping the name of a filterable diagnostic rule should trigger a warning. But at the same time, many different tools would like to use WGSL's diagnostic filter syntax for their own checks, so it should be possible for tools to ignore other tools' diagnostic names.

Although a warning does not prevent a shader from running, many developers adopt warning-free execution as a coding standard. WGSL should not prevent such developers from comfortably using third-party linter diagnostic names in their shaders.

How can WGSL warn about mistyped names, when the set of names must be open-ended?

Rust has diagnostic filtering very similar to WGSL's. For example, in Naga's source code, we say:

    #[allow(missing_docs)] // The names are self evident
    pub enum ShaderStage {
        Vertex,
        Fragment,
        Compute,
    }

If a diagnostic name like `missing_docs` is mistyped, Rust prints a warning.

Rust accommodates third-party diagnostics by allowing a prefix on diagnostic names. For example, from the `time` crate's source code:

    #![deny(
        anonymous_parameters,
        clippy::all,
        clippy::obfuscated_if_else,
        clippy::std_instead_of_core,
        clippy::undocumented_unsafe_blocks,
        const_err,
        illegal_floating_point_literal_pattern,
        late_bound_lifetime_arguments,
        rust_2018_idioms,
        rustdoc::broken_intra_doc_links,
        rustdoc::private_intra_doc_links
    )]

Here, `clippy` is a third-party Rust linter, and `rustdoc` is Rust's inline documentation processing tool, which is run separately from the Rust compiler.

The rule is simple: the Rust compiler warns if it sees an *unprefixed* name it doesn't recognize, and ignores all prefixed names. The rest is left to convention. So in the example above, Rust itself does not check prefixed diagnostic names like `rustdoc::broken_intra_doc_links` (but `rustdoc` will check that one).

In Rust, these are just names with a bit of lexical structure, not anything fancy. They're not affected by whatever definitions might be in scope for `clippy`, or `obfuscated_if_else`, or anything else. There's no connection between the `clippy::` prefix and any other definitions in the source code, or in the build configuration. This does imply that if a prefix is mistyped, like `cloppy::obfuscated_if_else`, nothing can warn about that.

WGSL should adapt Rust's approach as follows:

-   Expand the `diagnostic_rule_name` production as follows:

        diagnostic_rule_name:
          | ident_pattern_token
          | ident_pattern_token '.' ident_pattern_token
          
-   Replace the following paragraph in the specification:

    > An implementation may support filtering triggering rules not specified here, provided they are spelled differently from the standard triggering rules. A diagnostic filter with an unrecognized triggering rule will be ignored.
    
    with these:
    
    > If a diagnostic filter's triggering rule is unrecognized, then:
    >
    > -   If the rule name is a single identifier, a warning is triggered.
    >
    > -   If the rule name is a pair of identifiers separated by a '.',
    >     the diagnostic filter is silently ignored. 
    >
    > An implementation may support filtering triggering rules not specified here, provided their names are a pair of identifiers. Each implementation and WGSL processing tool should choose some identifier to appear consistently as the first part of the name for all non-standard diagnostics it implements.

Thus, one might write range diagnostic filters like this:

    @diagnostic(off, derivative_uniformity)
    @diagnostic(on, chromium.unreachable_code)
    fn function_that_should_have_no_unreachable_code_but_may_compute_derivatives_with_reckless_abandon() {
        ...
    }

Advantages of this approach:

- Typos get caught.
- This is better than a naming convention (say, `chromium_unreachable_code`) because it gives implementations a clear rule for distinguishing between diagnostic rule names they should check, and those they should not.
- It establishes an easy pattern that linters and other WGSL processors can adopt that allows their users to leave their diagnostic filters in the code without provoking warnings from browsers.
- It is not novel: Rust has been using this approach for years with good success.
- Although it is more complicated than simply ignoring unrecognized directives, it's not too baroque.

Disadvantages:

- It is more complicated than simply ignoring unrecognized directives.
- There is no way to detect typos in prefixes.

(For the sharp-eyed: The `#![...]` syntax in the second Rust example is Rust's analogue of a WGSL global diagnostic filter: it applies to the entire crate, not just the item that follows it.)

**Comment by ben-clayton** [comment]
> One potential advantage of this is that it would perhaps be clearer that these are not resolved in the same way that (future) namespaces are.

This is my main concern with @jimblandy 's proposal. `ident` `.` `ident` is a potential syntax for a namespace accessor, and if a user were to declare a `chromium` namespace in their WGSL, then it would certainly look like you are referencing a symbol in that namespace. The 3-arg `@diagnostic` is essentially a one-character variant on the original proposal, but avoids our concerns with accessor overloading.

**Comment by jimblandy** [comment]
I have no strong preference. Multi-arg has the same essential characteristics. But for what it's worth, based on my experience with this in Rust, people just don't think diagnostic names are identifiers for things in their program, so they don't confuse `clippy::` and `rustdoc::` with modules in their own code.

**Comment by jimblandy** [comment]
That proposal works for me.

I still prefer the dotted form. Rust's design has been tried in production and worked well. But I don't have any new arguments, beyond what I posted above.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-04-04-minutes#filterable-diagnostic-names-should-be-checked-but-also-permit-third-party-diagnostics--issue-3989--gpuwebgpuweb--github">WGSL 2023-04-04 Minutes</a></summary>

* JB: Looks like we have a consensus proposal.  Mozilla is fine with the three-arg form.  One final pitch for the dotted form; it takes the possessive notation from the language and uses it for this other thing.  That’s what Rust has done.  Rust has a very close facility.  It’s a tried and trusted form of solving this.  A minor notational question is much less important to us than giving warnings when the diagnostic name is mis-spelled.  It’s up to the committee.
* DN: As to having a preference, it’s probably BC who has the most concern about anticipating the design of modules, so maybe best to get him on the call before deciding.
* JB: If I’m the only one who likes the dots, I’ll cede, but also in my mind the similarity of dots to modules-ish things is a *benefit*, because that’s what we’re trying to draw to mind.
* DG: I also lean towards JB’s dots for similar reasons.
* DN: I imagine BC would ask: If we do modules, would we do identifier resolution lookup here?
* JB: No, just for suggestive effect.
* DN: Ok! **I’d rather have BC on the call then to move forward**.
* JB: Alright!

</details>

**Comment by ben-clayton** [comment]
@jimblandy and I chatted about this a bit in the office hours.

I worry that using a namespace syntax for browser-extension diagnostics (e.g. `@diagnostic(off, browser_namespace.diagnostic_name)`) is going to cause unnecessary (additional) confusion for our users.

In the spec today, we have the following special carve-outs for `name` in `@diagnostic(off, name)`:
* [`2.2.1. Filterable Triggering Rules`](https://www.w3.org/TR/WGSL/#filterable-triggering-rules): An implementation may support filtering triggering rules not specified here, provided they are spelled differently from the standard triggering rules. A [diagnostic filter](https://www.w3.org/TR/WGSL/#diagnostic-filter) with an unrecognized [triggering rule](https://www.w3.org/TR/WGSL/#diagnostic-triggering-rule) [will](https://www.w3.org/TR/WGSL/#behavioral-requirements) be ignored.
* [`3.9. Diagnostic Rule Names`](https://www.w3.org/TR/WGSL/#diagnostic-rule-names): The spelling of the token may be the same as an [identifier](https://www.w3.org/TR/WGSL/#identifier) but does not [resolve to](https://www.w3.org/TR/WGSL/#resolves) a declared object.

In layman's terms - `name` (and `off`!) in `@diagnostic(off, name)` is utterly special, and does not obey the regular, unified symbol resolution we've pushed for elsewhere in the language, and used by every other attribute. If I remember correctly, this was because of two reasons:
* A unresolved symbol should not be treated as an error, and so we can't use the standard identifier resolving rules.
* The [global diagnostic directive](https://www.w3.org/TR/WGSL/#global-diagnostic-filter) should be able to be processed before other module-scope declarations.

Example:
```ts
const derivative_uniformity = 42;
@diagnostic(off, derivative_uniformity) // derivative_uniformity does not resolve to the const above!
fn f() { /* ... */ }
```

Special rule carve-outs like this are unfortunate, but at the time, having the diagnostic-name identifier act differently seemed like an acceptable compromise.

With this proposal we're expanding the special carve-out to something that _looks_ like (the to-be-proposed) namespace member accessor expression, but in fact is not.

I'm particularly concerned about readability the day we allow users to create their own namespaces. Consider someone unfamiliar with the specialness of `@diagnostic`, who goes hunting for `unreachable_code` in the user-declared `browser_a` namespace:

```ts
namespace browser_a {
   const workgroup_size = 64; // A very precisely tuned number.
}
namespace browser_b {
   const workgroup_size = 128; // Another very precisely tuned number.
}

@compute @workgroup_size(${BROWSER}.workgroup_size)
@diagnosic(error, browser_a.unreachable_code)
fn main() { /* ... *./ }
```

In my humble opinion, this would be pushing the language (further) in a direction that requires the user to understand special-cases, with an increased risk of confusion to newcomers.

I feel the 3-parameter diagnostic is preferable to something that looks like a namespace-member expression.

After typing all of this, I'm now wishing we used strings, which would have avoided all the special rules all together:

```ts
// Uses `.` notation, but there's nothing to suggest that this does regular symbol resolution. Win-win!
@diagnostic("error", "browser_a.unreachable_code") 
```

**Comment by dneto0** [comment]
Agreement from the 2023-04-11 is to accept @jimblandy's proposal.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-04-11-minutes#filterable-diagnostic-names-should-be-checked-but-also-permit-third-party-diagnostics--issue-3989--gpuwebgpuweb--github">WGSL 2023-04-11 Minutes</a></summary>

* JB: I’d like to close this issue in this meeting. We have consensus on the most important things. Have one last question of the microsyntax rule. Feel like we’ve made our arguments clear.  Mine is using the pun of the namespace in the language is a production-tested design that has shipped and has proved helpful, and has not caused remarkable confusion.  Look for opinions from other members.
* BC: I dumped my thoughts in the issue.  Most of my crankiness is that the diagnostic attribute now is very special. It’s an inconsistency; it stands out vividly vs. all our other work to make things identifier-lookups.  That is upsetting to me on a fundamental level. If we go down member accesses as a tactic. When we add namespaces, it raises … I wrote a (contrived) example.  Think it’s minor.  In reflection, would have liked a string notation for the name.
* JB: So if we take the pun now, it would be weird in future.
* BC: My objection is we have things resolve differently in this one weird place. It’s inconsistent. It will look more weird by having something that looks like an expression that resolves differently
* KG: From an author’s standpoint it doesn’t seem too mystifying.  Understand from an implementors/ language lawyers standpoint.  If you asked an author to write it as a string literal, then they might say “why can’t you just know that and put the string marks in for me”.
* BC: May be my unfamiliar with Rust practice.  For someone unfamiliar with WGSL, how would they know it’s not resolved like other identifiers.
* DS: In my Rust experience I see a thing I don’t understand (‘diagnostic’) and then I go an look it up, and *that* tells me what it is and I’m ok.
* KG: Right, it is an inconsistency but tolerable.
* BC: Would lean against making it look like an expression. Weak preference. Won’t fight it.
* DN: My weak preference is tipped by Rust’s existing practice. We can reference that as prior practice.
* MM: Remind me what the feature is (rather than spelling it).
* go/
* MM: Concretely, e.g. spaces around equal sign.  Would have to be `webkit.spaces_around_equals`.  Because it’s not standardized.
* MM: Don’t want to stand in the way, but the trend has been the opposite. Trend has been to remove browser prefixes.
* JB: CSS is constrained by the practice to make sure …
* MM: Constrained is a little strong. 
* KG: Need a consensus.  Propose Jim’s consensus.
* BC: **Let’s go with Jim’s.**
* _Thanks everyone._

</details>
